File: VividFlow/CSIT321/static/js/SerializableObject.js
class SerializableObject {
    constructor() {
        this._ObjectID = "0.0"
        this.generate_object_id();
        this._Owner = null;
    }

    build_json_dict() {
        var json_dict = {};
        json_dict["ObjectType"] = SerialObjectTypes.get_friendly_name(this);
        json_dict["_ObjectID"] = this._ObjectID;
        return json_dict;
    }

    serialize_json() {
        return JSON.stringify(this.build_json_dict());
    }

    build_json_array(array) {
        var serialized_list = [];
        for ( let item of array) {
            serialized_list.push(item.build_json_dict());
        }
        return serialized_list;
    }

    reconstruct_from_json(jsonObj) {
        for (var prop in jsonObj) {
            if( this.hasOwnProperty(prop)) {
                if(Array.isArray(jsonObj[prop]))
                {
                    var tempArray = new Array();
                    for (var item in jsonObj[prop]) {
                        var newObj = SerialObjectTypes.build_object_from_json(jsonObj[prop][item]);
                        if (newObj != null)
                        {
                            tempArray.push(newObj);
                        }
                    }
                    this[prop] = tempArray;
                }
                else{
                    this[prop] = jsonObj[prop];
                }
            }
            else {
                if(prop == "ObjectType")
                {
                    //safe to ignore this property
                }
                else {
                    console.warn("JSON Reconstructor Warning: property '" + prop + "' does not exist on object of type '" + this.constructor.name + "'");
                }
            }
        }
    }

    get_object_id() {
        return this._ObjectID;
    }

    generate_object_id() {
        this._ObjectID = Math.random().toString();
    }

    set_owner(owner) {
        this._Owner = owner;
    }

    get_owner() {
        return this._Owner;
    }
}
SerialObjectTypes.register_type("SerializableObject", SerializableObject);

File: VividFlow/CSIT321/SerializableObject.py
import json
import Constants
import inspect
import random

class SerializableObject(object):
    _ObjectID = None

    def __init__(self):
        self.generate_object_id()

    def build_json_dict(self):
        json_dict = {}
        json_dict["ObjectType"] = Constants.SerialObjectTypes.get_friendly_name(self)
        json_dict["_ObjectID"] = self._ObjectID
        return json_dict

    def get_object_id(self):
        return self._ObjectID

    def generate_object_id(self):
        self._ObjectID = str(random.random())

    def serialize_json(self):
        json_dict = self.build_json_dict()
        json_str = json.dumps(json_dict)
        return json_str

    def build_json_array(self, array):
        serialized_list = []
        for item in array:
            serialized_list.append(item.build_json_dict())
        return serialized_list

    def __get_member_variables(self):
        attributes = inspect.getmembers(self, lambda a: not (inspect.isroutine(a)))
        members = [a[0] for a in attributes if not (a[0].startswith('__') and a[0].endswith('__'))]
        return members

    def reconstruct_from_json(self, json_obj):
        attributes = self.__get_member_variables()
        for key in json_obj:
            if key in attributes:
                if isinstance(json_obj[key], list):
                    tempList = []
                    for item in json_obj[key]:
                        new_obj = Constants.SerialObjectTypes.build_object_from_json(item)
                        if new_obj is not None:
                            tempList.append(new_obj)
                    self.__setattr__(key, tempList)
                else:
                    self.__setattr__(key, json_obj[key])
            else:
                if key == "ObjectType":
                    # safe to ignore this property
                    pass
                else:
                    print "JSON Reconstructor Warning: property '" + key + "' does not exist on object of type '" + type(self).__name__ + "'"
        pass

Constants.SerialObjectTypes.register_type("SerializableObject", SerializableObject)
